app/code/Magento/InventorySourceSelection/Model/DistanceProvider/GoogleMap/GetLatLngRequestFromAddress.php

```diff
@@ -0,0 +1,103 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\InventorySourceSelection\Model\DistanceProvider\GoogleMap;
+
+use Magento\Framework\Exception\LocalizedException;
+use Magento\Framework\HTTP\ClientInterface;
+use Magento\Framework\Serialize\Serializer\Json;
+use Magento\InventorySourceSelection\Model\DistanceProvider\GetLatLngRequestFromAddressInterface;
+use Magento\InventorySourceSelectionApi\Model\Request\LatLngRequestInterface;
+use Magento\InventorySourceSelectionApi\Model\Request\LatLngRequestInterfaceFactory;
+use Magento\InventorySourceSelectionApi\Api\Data\AddressRequestInterface;
+
+/**
+ * @inheritdoc
+ */
+class GetLatLngRequestFromAddress implements GetLatLngRequestFromAddressInterface
+{
+    public const GOOGLE_ENDPOINT = 'https://maps.google.com/maps/api/geocode/json';
+
+    /**
+     * @var array
+     */
+    private $latLngCache = [];
+
+    /**
+     * @var ClientInterface
+     */
+    private $client;
+
+    /**
+     * @var LatLngRequestInterfaceFactory
+     */
+    private $latLngRequestInterfaceFactory;
+
+    /**
+     * @var Json
+     */
+    private $json;
+
+    /**
+     * @var GetApiKey
+     */
+    private $getApiKey;
+
+    /**
+     * GetLatLngRequestFromAddress constructor.
+     *
+     * @param ClientInterface $client
+     * @param LatLngRequestInterfaceFactory $latLngRequestInterfaceFactory
+     * @param Json $json
+     * @param GetApiKey $getApiKey
+     */
+    public function __construct(
+        ClientInterface $client,
+        LatLngRequestInterfaceFactory $latLngRequestInterfaceFactory,
+        Json $json,
+        GetApiKey $getApiKey
+    ) {
+        $this->client = $client;
+        $this->latLngRequestInterfaceFactory = $latLngRequestInterfaceFactory;
+        $this->json = $json;
+        $this->getApiKey = $getApiKey;
+    }
+
+    /**
+     * @inheritdoc
+     * @throws LocalizedException
+     */
+    public function execute(AddressRequestInterface $addressRequest): LatLngRequestInterface
+    {
+        $cacheKey = $addressRequest->getAsString();
+        if (!isset($this->latLngCache[$cacheKey])) {
```

Comment:
why do we need this internal cache here?

Reply:
Because while iterating sources and addresses for distance comparison, this method can be called several times with the same address. Since the lat/lng is not supposed to change, I added a cache to avoid multiple API calls.